Prototyping of the Watchdog process

dmgav opened the pull request:
The first implementation of the Watchdog process that monitors the loop of RE Manager. In the new architecture RE Manager and RE Worker are started by the Watchdog process; RE Worker is started by request from RE Manager. Watchdog holds references to RE Worker and RE Manager. Data exchange between RE Worker and RE Manager is performed only using pipes.

The new command `kill_manager` is implemented in RE Manager and CLI tool (only for testing):
```
qserver -c kill_manager
```
This command blocks the event loop of RE Manager tripping the Watchdog, which restarts RE Manager in 5 seconds. After restart, RE Manager can communicate with the existing RE Worker (if the environment was created). The queue remains unchanged during the restart since it is stored in Redis. Currently restart will not alter operation off the worker when RE is idle, when a plan is running, when the plan is paused. The only known deficiency is that if execution of a plan is finished while manager is unresponsive, the execution of the next plan in the queue will not be started automatically.

danielballan commented on bluesky_queueserver/manager.py:
```diff
@@ -32,65 +29,107 @@
+class RunEngineManager(Process):
 
-    def __init__(self):
+    def __init__(self, *, conn_watchdog, conn_worker):
```
Should this extend rather than replace the argument accepted by `Process`, or else use composition rather than inheritance?

danielballan commented on bluesky_queueserver/manager.py:
```diff
@@ -329,6 +402,11 @@ async def _print_db_uids_handler(self, request):
+    async def _kill_manager_handler(self, request):
+        # This is expected to block the event loop forever
+        while True:
+            pass
```
Should there be a sleep in this loop to avoiding pinning a CPU?

dmgav replied:
Yes, sleep will work.

danielballan commented on bluesky_queueserver/manager.py:
```diff
@@ -387,10 +474,19 @@ async def zmq_server_comm(self):
+        try:
+            asyncio.run(self.zmq_server_comm())
+        except Exception as ex:
+            logger.info("Exiting RE Manager with exception %s" % str(ex))
```
The `logger.exception` method includes the whole traceback below the message.

```suggestion
            logger.exception("Exiting RE Manager")
```

dmgav replied:
Actually I didn't know about `logger.exception`. I will think about how to use it. Actually there is some work on logging that needs to be done.

danielballan commented on bluesky_queueserver/start_manager.py:
```diff
@@ -0,0 +1,162 @@
+                    # Messages should be handled in the event loop
+                    self._conn_received(msg)
+                except Exception as ex:
+                    logger.error("Exception occurred while waiting for packet: %s" % str(ex))
```
See note about `logger.exception`

danielballan commented on bluesky_queueserver/manager.py:
```diff
@@ -162,34 +203,53 @@ def _print_db_uids(self):
+                    # Messages should be handled in the event loop
+                    self._loop.call_soon_threadsafe(self._conn_watchdog_received, msg)
+                except Exception as ex:
+                    logger.error(f"Exception occurred while waiting for packet: {ex}")
```
See note about `logger.exception`

danielballan reviewed (commented):
Some comments for consideration. Feel free to ignore.

danielballan commented on bluesky_queueserver/start_manager.py:
```diff
@@ -0,0 +1,163 @@
+from .manager import RunEngineManager
+
+import logging
+logging.basicConfig(level=logging.DEBUG)
```
Now I understand why you were seeing overly-verbose log messages from ophyd. This configuration setting will add a handler that displays DEBUG-level messages from _all_ libraries. Library code should not configure handlers like this, at least not at import time. We did this in earlier versions of our libraries and learned to regret it. (Also see the gray "Note" box in the documentation section https://docs.python.org/3/howto/logging.html#configuring-logging-for-a-library for some official validation of that opinion.)

This could be added inside the `if __name__ == '__main__':` block (which is more an application than a library) and perhaps be configurable via a verbosity option like `-v`, as we do [in caproto](https://github.com/caproto/caproto/blob/4033c02da7815f7145c4a42581ddf0f76da4bbca/caproto/commandline/get.py#L140-L145).

danielballan commented on bluesky_queueserver/worker.py:
```diff
@@ -22,12 +18,14 @@
 
-config_bluesky_logging(level='INFO')
-config_ophyd_logging(level='INFO')
+logging.getLogger('bluesky').setLevel('WARNING')
```
This is a useful workaround in scripts or interactive sessions when loggers get to noisy, but modules should not silence loggers at import time. See previous comment for a better approach.

dmgav replied:
Controlling verbosity using -v is useful. I will add CLI options in one of the following PRs.

danielballan commented on bluesky_queueserver/manager.py:
```diff
@@ -3,12 +3,12 @@
+from .worker import DB
 
 import logging
 logging.basicConfig(level=logging.DEBUG)
```
This should be removed.

danielballan reviewed (changes requested):
The `master` branch uses `logging.basicConfig(level='DEBUG')` and should not have. That prompted workaround that doubled down on that approach, and it would be better to roll them back.

dmgav commented:
@danielballan Thank you for the comments. I guess I start to understand how it works now. Actually logging was waiting to be fixed in this project.

dmgav commented:
The last commit implements components necessary to restart RE Manager while plan is running in the worker process without disturbing the plan execution.

dmgav commented:
Moved code for RE Manager and RE Server to two separate directories. Created entry point `start-re-manager` to start RE manager from command line. Updated README.rst

dmgav commented:
The first 'real' test works with Python 3.7 and 3.8.

mrakitin reviewed (approved):
Tested during a call under @dmgav's supervision :)
Discussed a few improvements, which will be implemented in follow-up PRs.

tacaswell commented on bluesky_queueserver/manager/manager.py:
```diff
@@ -0,0 +1,655 @@
+    args, kwargs
+        `args` and `kwargs` of the `multiprocessing.Process`
+    """
+    def __init__(self, *args, conn_watchdog=None, conn_worker=None, **kwargs):
```
```suggestion
    def __init__(self, *args, conn_watchdog, conn_worker, **kwargs):
```

This will make them required kwargs (py36 feature iirc).

tacaswell commented on bluesky_queueserver/manager/manager.py:
```diff
@@ -0,0 +1,655 @@
+        Names of plans and devices are strings.
+        """
+        n_pending_plans = await self._r_pool.llen('plan_queue')
+        logger.info("Starting a new plan: %d plans are left in the queue" % n_pending_plans)
```
```suggestion
        logger.info("Starting a new plan: %d plans are left in the queue", n_pending_plans)
```

This defers doing the templating until the loggers need it.

dmgav replied:
I see, I will modify the logging commands in the next PR.

tacaswell commented:
I'll continue my review as a PR ;)
